Samples/DirectMLNpuInference/main.cpp

```diff
@@ -13,11 +13,36 @@
 
 using Microsoft::WRL::ComPtr;
 
-void InitializeDirectML(ID3D12Device1** d3dDeviceOut, ID3D12CommandQueue** commandQueueOut, IDMLDevice** dmlDeviceOut) {
-    // Whether to skip adapters which support Graphics in order to target NPU for testing
-    bool forceComputeOnlyDevice = true;
-    bool forceGenericMLDevice = false;
-    
+void InitializeDirectML(ID3D12Device1** d3dDeviceOut, ID3D12CommandQueue** commandQueueOut, IDMLDevice** dmlDeviceOut)
+{
+    // Useful Flags to chage to allow, require, and disallow certain attributes of devices for testing.
+    const bool allowGraphicsCapabilities = false;
+    const bool requireComputeDevice = true;
+    const bool requireGenericMLDevice = false;
+
+    // Populate helper structures based on above flags
+    std::vector<GUID> dxGuidAllowedAttributes = {};
+    std::vector<GUID> dxGuidRequiredAttributes = {};
```

Comment:
Do we need both allowed and required, if we are going to filter out anything in allowed that's not in required?

Reply:
The allow list is not technically needed, but due to the way that CreateAdapterList requires minimum one attribute and filters for all attributes, we do need a default list of attributes to individually loop over, when no required attributes were specified.

There isn't a lot that could be done to make the code look simpler here that I can think of, but I'll add a condition that iterates over the required list instead of the allowed list if the require list is non-empty

Reply:
Would it be a bit easier to do the following, only with required attributes? This way, we don't need dxGuidAllowedAttributes or do additional filtering later?

if (requireComputeDevice)
{
 dxGuidRequiredAttributes.push_back(DXCORE_ADAPTER_ATTRIBUTE_D3D12_CORE_COMPUTE);
}

if (requireGenericMLDevice)
{
 dxGuidRequiredAttributes.push_back(DXCORE_ADAPTER_ATTRIBUTE_D3D12_GENERIC_ML);
}

if (dxGuidRequiredAttributes.empty())
{
 dxGuidRequiredAttributes.push_back(DXCORE_ADAPTER_ATTRIBUTE_D3D12_CORE_COMPUTE);
 dxGuidRequiredAttributes.push_back(DXCORE_ADAPTER_ATTRIBUTE_D3D12_GENERIC_ML);
}

Reply:
A challenge with this way is that dxGuidRequiredAttributes can't distinguish between "all attributes in this list are required" if both require flags are true and "any attribute in this list is ok" if both require flags are empty.

My current mental model of the require flags e.g. requireGenericMLDevice is that enabling multiple flags means that all of those attributes should be supported